File: Top_100_Liked_Questions/309._Best_Time_to_Buy_and_Sell_Stock_with_Cooldown/solution.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l = len(prices)
        mem = [[0]*(l+1) for _ in range(l+2)]
        
        for i in range(l-1, -1, -1):
            mem[i][0] = max(mem[i+1][i+1], mem[i+1][0])
            for j in range(i):
                mem[i][j+1] = max(prices[i] - prices[j] + mem[i+2][0], mem[i+1][j+1])
        
        return mem[0][0]
        
    
    def maxProfit2(self, prices: List[int]) -> int:
        # recursive dp
        def helper(idx, bought_idx):
            if idx >= len(prices):
                return 0;
            
            if bought_idx == -1:
                return max(helper(idx+1, idx), helper(idx+1, -1))
            
            return max(prices[idx] - prices[bought_idx] + helper(idx+2, -1), helper(idx+1, bought_idx))
    
        return helper(0, -1)

File: Top_100_Liked_Questions/309._Best_Time_to_Buy_and_Sell_Stock_with_Cooldown/solution.java
class Solution {
    // iterative dp
    public int maxProfit(int[] prices) {
        int[][] mem = new int[prices.length+2][prices.length+1];
        for (int i=prices.length-1; i>-1; i--) {
            mem[i][0] = Math.max(mem[i+1][i+1], mem[i+1][0]);
            for (int j=0; j<i; j++) {
                mem[i][j+1] = Math.max(prices[i] - prices[j] + mem[i+2][0], mem[i+1][j+1]);
            }
        }
        return mem[0][0];
    }
}


class Solution2 {
    // recursive dp
    public int maxProfit(int[] prices) {
        return this.maxProfit(prices, 0, -1);
    }

    public int maxProfit(int[] prices, int idx, int buy_idx) {
        if (idx >= prices.length)
            return 0;

        if (buy_idx == -1)
            return Math.max(this.maxProfit(prices, idx+1, idx), this.maxProfit(prices, idx+1, -1));

        return Math.max(prices[idx] - prices[buy_idx] + this.maxProfit(prices, idx+2, -1), this.maxProfit(prices, idx+1, buy_idx));
    }
}